File: slidingLogs.go
// RateLimitUsingSlidingLogs .
func RateLimitUsingSlidingLogs(userID string, uniqueRequestID string, intervalInSeconds int64, maximumRequests int64) bool {
	// userID can be apikey, location, ip
	currentTime := strconv.FormatInt(time.Now().Unix(), 10)
	lastWindowTime := strconv.FormatInt(time.Now().Unix()-intervalInSeconds, 10)
	// get current window count
	requestCount := redisClient.ZCount(ctx, userID, lastWindowTime, currentTime).Val()
	if requestCount >= maximumRequests {
		// drop request
		return false
	}

	// add request id to last window
	redisClient.ZAdd(ctx, userID, &redis.Z{Score: float64(time.Now().Unix()), Member: uniqueRequestID}) // add unique request id to userID set with score as current time

	// handle request
	return true
	// remove all expired request ids at regular intervals using using ZRemRangeByScore from -inf to last window time
}

File: slidingLogs.ts
import { RedisClientType, createClient, ZAddOptions } from "redis";

type RedisClient = RedisClientType;

const redisClient: RedisClient = createClient();

interface RateLimitOptions {
  userID: string;
  uniqueRequestID: string;
  intervalInSeconds: number;
  maximumRequests: number;
}

async function rateLimitUsingSlidingLogs({
  userID,
  uniqueRequestID,
  intervalInSeconds,
  maximumRequests,
}: RateLimitOptions): Promise<boolean> {
  const currentTime = Math.floor(Date.now() / 1000).toString();
  const lastWindowTime = (
    Math.floor(Date.now() / 1000) - intervalInSeconds
  ).toString();

  // Get current window count
  const requestCount = await redisClient.zCount(
    userID,
    lastWindowTime,
    currentTime
  );

  if (requestCount >= maximumRequests) {
    // Drop request
    return false;
  }

  // Add request id to last window
  const zAddOptions: ZAddOptions = {
    score: Date.now() / 1000,
    value: uniqueRequestID,
  };
  await redisClient.zAdd(userID, zAddOptions);

  // Handle request
  return true;

  // Optionally, you can remove all expired request ids at regular intervals using ZRemRangeByScore
  // await redisClient.zRemRangeByScore(userID, '-inf', lastWindowTime);
}

export { rateLimitUsingSlidingLogs };
